lib/action_dispatch/session/active_record_store.rb

```diff
@@ -77,9 +77,25 @@ def get_session(request, sid)
           [sid, session.data]
         end
       end
+      
+      def extract_session_id(request)
+        stale_session_check! do
+          unpacked_cookie_data(request)
+        end
+      end
+      
+      def unpacked_cookie_data(request)
+        request.fetch_header("action_dispatch.request.unsigned_session_cookie") do |k|
+          v = stale_session_check! do
+            get_cookie(request) || {}
+          end
+          request.set_header k, v
+        end
+      end
 
       def write_session(request, sid, session_data, options)
         logger.silence_logger do
+          sid ||= get_cookie(request)
```

Comment:
Why do we need to read the sid from the cookie? Is not it already passed to `write_session`?